### PR TITLE
Add OpenXR fb_render_model extension wrapper and OpenXRFBRenderModel node

### PR DESCRIPTION
**Pull request by devloglogan**
Adds an extension wrapper for the OpenXR [XR_FB_render_model](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_FB_render_model) extension and corresponding node implementing its usage. This extension provides access to gltf models of controllers/keyboards at runtime.

The extension wrapper caches a list of valid paths to these models (accessible with function `get_valid_paths()` in the `OpenXRFBRenderModel` node ~~or the same function in the extension wrapper singleton~~). One of these paths can be assigned to the `render_model_path` property of the `OpenXRFBRenderModel` node, which will trigger that model to be loaded and accessible via the nodes `get_render_model()` function. The only other time the model is loaded is on startup via the `OpenXrInterface` `session_begun` signal.

At present (tested on Quest 2 and Quest 3), it seems only the controller models are able to be successfully loaded. The virtual keyboard requires the XR_META_virtual_keyboard extension to be enabled, but I believe the other two _should_ be available. Unsure if there's another dependency missing for these. The controller models have been added to the demo main scene.

### CODE REVIEW

**Review by m4gr3d — CHANGES_REQUESTED** [review]
Looks great! I've added some minor feedback to address.

File: demo/main.gd
Comment (devloglogan) [reply demo/main.gd:37]: It re-positions the hand tablet child node to the correct spot for the grip pose.

Comment (dsnopek) [reply demo/main.gd:37]: I think we could just make these changes permanently to the scene (except the visibility change, which needs to be in code)

File: godotopenxrmeta/src/main/cpp/register_types.cpp
Comment (dsnopek) [reply godotopenxrmeta/src/main/cpp/register_types.cpp:87]: Because it's a `Node`, I think it needs to be registered in `SCENE`

Comment (devloglogan) [reply godotopenxrmeta/src/main/cpp/register_types.cpp:87]: Initially that's what I tried, but the node isn't available in the editor if registered there. @dsnopek informed me that the scene level is the correct choice here.

File: godotopenxrmeta/src/main/cpp/openxr_fb_render_model_extension_wrapper.cpp
Comment (devloglogan) [reply godotopenxrmeta/src/main/cpp/openxr_fb_render_model_extension_wrapper.cpp:65]: Whoops, definitely meant to remove `get_buffer()` from here. And it does feel a bit redundant to have `get_valid_paths()` available in both. I'll remove these.

File: godotopenxrmeta/src/main/cpp/openxr_fb_render_model_extension_wrapper.cpp
Comment (devloglogan) [reply godotopenxrmeta/src/main/cpp/openxr_fb_render_model_extension_wrapper.cpp:115]: Good to know, thank you!

File: godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp
Comment (Malcolmnixon) [inline godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp:118]: Is it possible to help the user enter a valid model path here. A similar issue arises with XRNode3D where there's a pose property string, but the list of poses cannot be statically entered (as it comes from the OpenXR Action Map) and so must be queried at runtime.

XRNode3D implements this by adding the `PROPERTY_HINT_ENUM_SUGGESTION` option as:
https://github.com/godotengine/godot/blob/74c32faa78b54863f8f25c538083907c2bf71791/scene/3d/xr_nodes.cpp#L230
This changes the node-editor to show a drop-down, but still lets the developer click a pen icon and type in a custom string should they choose to.

Then it implements the `_validate_property` method and uses it to populate the hint-strings from the `xr_server->get_suggested_pose_names`:
https://github.com/godotengine/godot/blob/74c32faa78b54863f8f25c538083907c2bf71791/scene/3d/xr_nodes.cpp#L240-L259
You should be able to use the `get_valid_paths` method to populate valid paths.

**Review by Malcolmnixon — APPROVED** [review]
This works well for me!

File: godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp
Comment (dsnopek) [reply godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp:118]: Oh, that's a great idea! I actually didn't realize you could do a dropdown as a suggestion.

However, I don't think using `get_valid_paths()` would work - it'll only return results when we have a running OpenXR session, and that'll only be true in the XR editor. But we could certainly provide all the paths in the spec as suggestions!

Comment (BastiaanOlij) [reply godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp:118]: Worse yet, even if we get it to run in the editor it won't help you, because you'll get a list of valid paths on the PC. On Windows you might get lucky if you've got the Oculus Client configured as OpenXR, but you're dead in the water on Mac.

This is why in the action map our meta data ignores extensions, we can configure the action map regardless of the capabilities of the system on which the user is developing their game. What is important is what the capabilities of the headset are the user is running this on.

**Comment by BastiaanOlij** [comment]
Looks really good at an early glance.

I'm concerned about the requirement of the rotation of the grip pose to position the render model, that should match up or the spec should provide information about correct placement. If it doesn't that's going to be bad and bite us down the road if other vendors follow suit but do properly position the meshes.

On that subject, exposing the render models to the end user should take a more global approach and not be FB centric. Here I'm unfortunately not at liberty to provide more detail yet, but we're going to back ourselves into a corner if we do not take a wider view.

**Comment by devloglogan** [comment]
> I'm concerned about the requirement of the rotation of the grip pose to position the render model

To be clear, the render model isn't being repositioned by anything other than the setting of the grip pose. The hacky rotation is only there for repositioning the hand tablet correctly from one pose to the other. Would this not be expected? We could just set the pose for the controllers to be grip by default and reposition the hand tablet in the editor instead.

**Comment by dsnopek** [comment]
> We could just set the pose for the controllers to be grip by default and reposition the hand tablet in the editor instead.

Yes, I think we should do exactly this!

**Comment by devloglogan** [comment]
Alright, I changed the default controller poses to grip and added some (hard coded) `render_model_path` suggestions

**Comment by dsnopek** [comment]
> On that subject, exposing the render models to the user should take a more global approach and not be FB centric. Here I'm unfortunately not at liberty to provide more detail yet, but we're going to back ourselves into a corner if we do not take a wider view.

If a vendor neutral OpenXR extension is added to the spec, we'd implement that in Godot itself'. So, personally, I don't think having the Meta-specific one here in 'godot_openxr_vendors' paints us into any corners. Folks not specifically targeting Meta headsets won't use their vendor extensions (they'll wait for the vendor neutral one), and Meta's extension will keep working on Meta devices even after the vendor neutral one exists.

File: godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp
Comment (m4gr3d) [inline godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp:63]: We should omit the keyboard paths until we add support for the keyboard extension. Maybe add a todo pointing to an issue to track.

**Review by m4gr3d — APPROVED** [review]
One minor comment to address but otherwise it looks good!

File: demo/main.gd
Comment (dsnopek) [inline demo/main.gd:37]: I think this should just be part of the scene too. There is no reason for the tablet to be a child of what it's a child of currently, so let's just move it in the editor permanently. Otherwise, folks may look at this demo thinking there's some reason for doing this, and copy it into their own projects :-)

File: demo/main.tscn
Comment (dsnopek) [inline demo/main.tscn:72]: Why not just have this as a child of `LeftHand`? Ie

```suggestion
[node name="Viewport2Din3D" parent="XROrigin3D/LeftHand/HandTablet" instance=ExtResource("2_7whgo")]
```

Comment (devloglogan) [reply demo/main.tscn]: Good point, moved it!

**Comment by BastiaanOlij** [comment]
Just an intermediate update from me as I'm waiting on hearing back before we can make an informed decision on how to proceed here. Until I can share updates, best to press pause, hope to have info in the next couple days.

**Comment by BastiaanOlij** [comment]
> If a vendor neutral OpenXR extension is added to the spec, we'd implement that in Godot itself. So, personally, I don't think having the Meta-specific one here in 'godot_openxr_vendors' paints us into any corners. Folks not specifically targeting Meta headsets won't use their vendor extensions (they'll wait for the vendor neutral one), and Meta's extension will keep working on Meta devices even after the vendor neutral one exists.

It's not that simple. The following scenario is fairly likely:
- we implement the meta extension as is
- in the near future we implement the official solution but there is no guarantee Meta will immediately support it as they already have a solution
- our users need to build something that uses the Meta solution on Quest, and the official solution on everything else, the most likely scenario is simply checking which one is available.
- Meta adds supports for the official solution but keeps the Meta solution around for backwards compatibility, both are now reported as available
- Existing Godot apps potentially will break as they try to activate both Meta and official solutions at the same time.

So we have a mandate to at least discuss whether we can come up with a solution that staves off this scenario (which is very likely due to reasons I'm not at liberty to share), or whether we should forge ahead and deal with the fallout later.

File: godotopenxrmeta/src/main/cpp/include/openxr_fb_render_model.h
Comment (m4gr3d) [inline godotopenxrmeta/src/main/cpp/include/openxr_fb_render_model.h:59]: Let's remove this method. Through the api users should only have access to what we know is supported by the runtime.

Comment (dsnopek) [reply godotopenxrmeta/src/main/cpp/include/openxr_fb_render_model.h:59]: What do you mean by this? The user needs some way to select which model should be loaded. Do you mean switching from a `String` to an `enum` so they can only use certain values, rather than a free-form string?

**Comment by devloglogan** [comment]
@BastiaanOlij rebased!

File: godotopenxrmeta/src/main/cpp/include/openxr_fb_render_model.h
Comment (m4gr3d) [reply godotopenxrmeta/src/main/cpp/include/openxr_fb_render_model.h]: Yeah I was referring to remove users' ability to enter a path themselves, so switching to an `enum` would work.

**Comment by m4gr3d** [comment]
@BastiaanOlij @dsnopek From our discussions, it's likely we'll move forward with this approach as a `v0` api. Can you review in that context.

@devloglogan Can you rebase the PR for review.

**Comment by devloglogan** [comment]
Rebased! Per Fredia's request I've swapped the model selector on the `OpenXRFbRenderModel` node to be an `enum` type. After doing that, it seemed pointless storing path names and exposing them, so I removed that functionality.

File: common/src/main/cpp/extensions/openxr_fb_render_model.cpp
Comment (m4gr3d) [inline common/src/main/cpp/extensions/openxr_fb_render_model.cpp:2]: Can you move this implementation into a `classes` directory: `common/src/main/cpp/classes`. 

This will enable us to distinguish between extension wrappers implementations in the `extensions` directory and nodes / classes used to expose the functionality to end users.

File: common/src/main/cpp/extensions/openxr_fb_render_model.cpp
Comment (m4gr3d) [inline common/src/main/cpp/extensions/openxr_fb_render_model.cpp:130]: Let's make the signal name more explicit and use `openxr_fb_render_model_loaded`.

File: common/src/main/cpp/include/extensions/openxr_fb_render_model.h
Comment (m4gr3d) [inline common/src/main/cpp/include/extensions/openxr_fb_render_model.h:2]: Similar comment as for `openxr_fb_render_model.cpp`; let's move this header under a new `include/classes` subdirectory: `common/src/main/cpp/include/classes`

**Review by m4gr3d — CHANGES_REQUESTED** [review]

**Comment by devloglogan** [comment]
@m4gr3d requests added!

**Review by dsnopek — APPROVED** [review]
This is looking great to me!

I wonder if the "classes" directory should really be "scene" because this is a node and that's what Godot names the directory that holds nodes? If we end up with some `RefCounted` classes it may be nice to have those in a different directory. But what we have here is certainly fine for now :-)

**Review by m4gr3d — APPROVED** [review]

File: CHANGES.md
Comment (m4gr3d) [inline CHANGES.md:7]: Typo in `wrapper`

Comment (devloglogan) [reply CHANGES.md:7]: Ahg, my bad, thank you!

**Review by BastiaanOlij — CHANGES_REQUESTED** [review]
Ok Logan, just starting. Getting everything to compile and to a point where I thought I could run things took a lot longer than expected, and I ran into a number of snags I'm still looking into. Part of it was me getting familiar with things again (haven't had a deep dive into the vendors plugin), part is that we may have lost some instructions or that a clean build is throwing up some issues.

So there is more to come. So far it looks good, I'm mostly worried about some flow things around the model loading at the wrong time and in the wrong scenario.

Also my apologies that it took so long before I could really sink my teeth into. There are things happening on the core of OpenXR where doing something a little more elaborate to make this a bit more future proof may have been a necessary thing.

But looking at some structural differences and the fact that things will take longer than I hoped, I'm now of the belief that this is too soon and that an FB specific implementation as you've done here, is the right course of action. I also do not believe it likely that other vendors will come out with their own extension which would require us to think this through a little more.

That will mean that in the near future users who implement this extension and are planning to release games beyond the Meta platform, will likely need to refactor their code. We will deal with that, when the time is right.

File: common/src/main/cpp/classes/openxr_fb_render_model.cpp
Comment (devloglogan) [reply common/src/main/cpp/classes/openxr_fb_render_model.cpp:44]: I've made some changes, hope they will suffice! There's three scenarios at which a model can now be loaded.

1. The `session_begun` signal, for the `OpenXRFbRenderModel` nodes that are added to the scene tree in the editor.

2. The `NOTIFICATION_ENTER_TREE` notification, which checks to see if the OpenXR session is active (I'm now storing that in the extension wrapper). For the `OpenXRFbRenderModel` nodes that are added at runtime.

3. The property setting route, which now checks that it's in the scene tree and that the session is active. For changing the model type at runtime.

File: common/src/main/cpp/include/classes/openxr_fb_render_model.h
Comment (devloglogan) [reply common/src/main/cpp/include/classes/openxr_fb_render_model.h]: Renamed them. Personally, I prefer the more verbose names as just `type` and `node` feel a bit ambiguous to me. But if you feel strongly on dropping the prefix just let me know!

File: common/src/main/cpp/classes/openxr_fb_render_model.cpp
Comment (BastiaanOlij) [reply common/src/main/cpp/classes/openxr_fb_render_model.cpp]: Looks good to me :)

**Comment by m4gr3d** [comment]
@BastiaanOlij Looks like all your feedback on the PR have been addressed. Are you good with merging the current implementation?

**Comment by BastiaanOlij** [comment]
@m4gr3d I haven't gotten anywhere near actually testing the latest version, as I was discussing with David the other day I am getting fairly concerned with this, especially as the testing I did get around doing last week show that a few things have snuck in that are causing issues and right now the plugin is in a broken state for me.

Purely based on the code and the discussion around this however, I am happy. But I would like to actually take it for a drive before giving my final thumbs up.

File: common/src/main/cpp/classes/openxr_fb_render_model.cpp
Comment (BastiaanOlij) [inline common/src/main/cpp/classes/openxr_fb_render_model.cpp:137]: I'm not sure about the best implementation but the one thing I was missing is control/signals related to the controller no longer being tracked.

Right now in the demo you react to the `openxr_fb_render_model_loaded` signal, and hide the hand tracking but that seems to be too restrictive, and once we use proper hand tracking, it should actually show the hand properly holding the controller (and thus verifying the positioning is correct).

Off course this could be a limitation of the demo, and possibly the logic I'm looking for shouldn't be part of this but should be part of the general hand tracking implementation that is in the core of Godot.

If the latter, we should probably do a follow up PR that adds hand tracking into the demo with the hand meshes Malcolm has (I plan to do a version of this during the week for the `godot-demo-projects` repo).

But if there is more to handle here, especially if tracking of controllers stops and the controllers should be hidden, that is worth looking into before we merge this.

**Review by BastiaanOlij — APPROVED** [review]
OK, so thanks to some motivational help from Fredia I managed to get things working today. There are some issues unrelated to this PR but the big issues that got me stuck and ready to throw in the proverbial towel, seemed to have been solved.

The only one that worries me but is likely again unrelated to this PR, is that the Godot IDE hung while I was looking around at the setup.

There is one question that should be down below for what happens when controllers are let go and some suggestions for improving the demo. I'm happy for this to be merged and that issue to be raised as a separate issue to be handled in a follow up PR.

File: common/src/main/cpp/classes/openxr_fb_render_model.cpp
Comment (devloglogan) [reply common/src/main/cpp/classes/openxr_fb_render_model.cpp:137]: My initial thought: this extension exists to provide 3D models, so our implementation should just expose those 3D models, and any further logic should be done by end users (or at least not here). It's possible users may want to drop static controller models in other areas, not just use them for the tracked controllers.

Comment (dsnopek) [reply common/src/main/cpp/classes/openxr_fb_render_model.cpp:137]: I agree with that. There's also more models than just controllers, this can also be used for physics keyboards (for the tracked keyboard feature) and the advanced virtual keyboard (which isn't tracking anything at all).

**Comment by m4gr3d** [comment]
@devloglogan Can you rebase the PR so we can proceed to merge it!

**Comment by devloglogan** [comment]
@m4gr3d rebased

**Pull request merged**
